grudge/op.py

```diff
@@ -47,38 +24,136 @@
 """
 
 
+from arraycontext.metadata import FirstAxisIsElementsTag
+from arraycontext.loopy import make_loopy_program
+from arraycontext.container.traversal import freeze
+
 from numbers import Number
-from pytools import memoize_on_first_arg
 
-import numpy as np  # noqa
+from functools import reduce
+
+from pytools import (
+    memoize_in,
+    memoize_on_first_arg,
+    keyed_memoize_in
+)
 from pytools.obj_array import obj_array_vectorize, make_obj_array
-import pyopencl.array as cla  # noqa
-from grudge import sym, bind
 
+from meshmode.dof_array import DOFArray
+
+import numpy as np
 import grudge.dof_desc as dof_desc
 
-from meshmode.mesh import BTAG_ALL, BTAG_NONE, BTAG_PARTITION  # noqa
-from meshmode.dof_array import freeze, flatten, unflatten
+from grudge.trace_pair import (  # noqa
+    interior_trace_pair,
+    interior_trace_pairs,
+    connected_ranks,
+    cross_rank_trace_pairs,
+    bdry_trace_pair,
+    bv_trace_pair
+)
+
+
+__doc__ = """
+Metadata and geometry
+^^^^^^^^^^^^^^^^^^^^^
 
-from grudge.symbolic.primitives import TracePair
+Kernel tags
+-----------
 
+.. autoclass:: HasElementwiseMatvecTag
+
+Projection and interpolation
+----------------------------
+
+.. autofunction:: project
+
+Geometric quantities
+--------------------
+
+.. autofunction:: nodes
+.. autofunction:: normal
+.. autofunction:: h_max_from_volume
+.. autofunction:: h_min_from_volume
+
+Core DG routines
+^^^^^^^^^^^^^^^^
+
+Elementwise differentiation
+---------------------------
+
+.. autofunction:: local_grad
+.. autofunction:: local_d_dx
+.. autofunction:: local_div
+
+Weak derivative operators
+-------------------------
+
+.. autofunction:: weak_local_grad
+.. autofunction:: weak_local_d_dx
+.. autofunction:: weak_local_div
+
+Mass, inverse mass, and face mass operators
+-------------------------------------------
+
+.. autofunction:: mass
+.. autofunction:: inverse_mass
+.. autofunction:: face_mass
+
+Support functions
+^^^^^^^^^^^^^^^^^
+
+Nodal reductions
+----------------
+
+.. autofunction:: norm
+.. autofunction:: nodal_summation
+.. autofunction:: nodal_minimum
+.. autofunction:: nodal_maximum
+.. autofunction:: integral
+
+Elementwise reductions
+----------------------
+
+.. autofunction:: elementwise_sum
+"""
 
+
+# {{{ Kernel tags
+
+class HasElementwiseMatvecTag(FirstAxisIsElementsTag):
+    """A tag that is applicable to kernel programs indicating that
+    an element-wise matrix product is being performed. This indicates
+    that the first index corresponds to element indices and suggests that
+    the implementation should set element indices as the outermost
+    loop extent.
+    """
+
+# }}}
```

Comment:
@inducer Should we move tags to something like `metadata.py`?

Reply:
Nah. Not yet... the transform story is still too up in the air.